// bema-utils/lib/bins/index.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});/**
 *
 * Bins Utility Plugin
 *
 * @author Aldair Beltran
 *
 */

const WHITE_LIST = ['https://hastebin.com', 'https://pastie.io'];

var _net = require('../net');

 class Bin {
    
	constructor(url) {
		this.url = url;
	}
	async upload(toUpload) {
        try {
            const chunk = await _net.UtilNetwork.call(void 0, this.url).request({
                path: '/documents',
                data: toUpload,
            }) ;
            try {
                const linkStr = this.url + '/' + JSON.parse(chunk.toString())['key'];
                return linkStr;
            } catch (err) {
                return err;
            }
        } catch (e) {
            return e;
        }
	}
	async download(key) {
        const url = this.url + key;
        try {
            return await _net.UtilNetwork.call(void 0, url).get();
        } catch (e) {
            return e;
        }
	}
} exports.Bin = Bin;

 const Pastie = new Bin(WHITE_LIST[1]); exports.Pastie = Pastie;
 const Hastebin = new Bin(WHITE_LIST[0]); exports.Hastebin = Hastebin;

 async function upload(toUpload) {
    try {
        return await exports.Hastebin.upload(toUpload);
    } catch (e) {
        try {
            return await exports.Pastie.upload(toUpload);
        } catch (err) {
            return err;
        }
    }
} exports.upload = upload;
 async function download(key) {
    try {
        return await exports.Hastebin.download(key);
    } catch (e) {
        try {
            return await exports.Pastie.download(key);
        } catch (err) {
            return err;
        }
    }
} exports.download = download;
