// bema-utils/lib/net/index.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});/***
 *
 * Network Utility functions
 *
 * @author Aldair Beltran
 *
 */

var _https = require('https'); var https = _https;
var _http = require('http'); var http = _http;
var _url = require('url'); var url = _url;

 const description = 'Easy way to do request from some http/s page'; exports.description = description;




 class Net {
	
	
	constructor(uri) {
		this.url = uri;
		this.protocol = url.parse(this.url).protocol ;
	}
	get() {
		const net = this.protocol === 'https:' ? https : http;
		return new Promise((resolve, reject) => {
			net.get(this.url, (res) => {
				res.setEncoding('utf8');
				let data = '';
				res.on('data', (chunk) => {
					data += chunk;
				});
				res.on('end', () => {
					resolve(data );
				});
				res.on('error', (err) => {
					reject(err);
				});
			})
				.on('error', (err) => {
					reject(err);
				})
				.setTimeout(3500);
		});
	}
	async toJSON() {
		try {
			const data = await this.get();
			return JSON.parse(data);
		} catch (e) {
			return e;
		}
	}
	request(opts) {
		const net = this.protocol === 'https:' ? https : http;
		const actionUrl = url.parse(this.url);
		const hostname = actionUrl.hostname;
		const options = {
			hostname: hostname,
			method: 'POST',
		};
		// eslint-disable-next-line
		for (let i in opts) {
			options[i] = opts[i];
		}
		if (options.data) delete options.data;
		if (!options.hostname) options.hostname = hostname;
		return new Promise((resolve, reject) => {
			let str = '';
			const req = net.request(options, (res) => {
				res.setEncoding('utf8');
				res.on('data', (chunk) => {
					str += chunk;
				});
				res.on('end', () => {
					resolve(str);
				});
			});
			req.on('error', (e) => {
				reject(e);
			});
			if (opts.data) req.write(opts.data);
			req.end();
		});
	}
} exports.Net = Net;
 function UtilNetwork(uri) {
	return new Net(uri);
} exports.UtilNetwork = UtilNetwork;
