// bema-utils/lib/package/index.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});
var _fs = require('../fs');




 class Package {
    
    
    constructor(file) {
        if (!file) {
            this.file = './';
        } else {
            this.file = `${file}${file.endsWith('/') ? '' : '/'}`;
		}
		this.file += 'package.json';
        this.data = {};
	}
	getData() {
		if (this.exists) {
			return JSON.parse(_fs.FS.call(void 0, this.file).readExistsSync());
		} else {
			return JSON.parse('{}');
		}
	}
	get name() {
		const pathName = this.file.split('/');
		return pathName[pathName.length - 2];
	}
    creator(data) {
        if (typeof data !== 'object') throw Error('Unacceptable data type');
		Object.assign(this.data, data);
		if (this.data.main) {
			delete this.data.main;
		}
		if (this.data.name) {
			delete this.data.name;
		}
		this.data.main = this.name;
		this.data.name = this.name;
		const toSave = JSON.stringify(this.data);
        _fs.FS.call(void 0, this.file).writeSync(toSave);
    }
    get exists() {
		try {
            return _fs.FS.call(void 0, this.file).isFileSync();
		} catch (e) {
			return false;
		}
    }
    remove() {
        try {
            _fs.FS.call(void 0, this.file).unlinkIfExistsSync();
            return true;
        } catch (e) {
            return false;
        }
    }
} exports.Package = Package;
